base/sort.jl

```diff
@@ -760,133 +1071,119 @@ function _issorted(v::AbstractVector, lo::Integer, hi::Integer, o::Ordering)
+
+We next apply [`BoolOptimization`](@ref) because it also has no runtime cost when it is not
+triggered and when it is triggered, it is an incredibly efficient algorithm (sorting `Bool`s
+is quite easy).
+
+Next, we dispatch to [`InsertionSort`](@ref) for inputs with `length <= 10`. This dispatch
+occurs before the [`IEEEFloatOptimization`](@ref) pass because the
+[`IEEEFloatOptimization`](@ref)s are not beneficial for very small inputs.
+
+To conclude the [`InitialOptimizations`](@ref), we apply [`IEEEFloatOptimization`](@ref).
+
+After these optimizations, we branch on whether radix sort and related algorithms can be
+applied to the input vector and ordering. We conduct this branch by testing if
+`UIntMappable(v, order) !== nothing`. That is, we see if we know of a reversible mapping
+from `eltype(v)` to `UInt` that preserves the ordering `order`. We perform this check after
+the initial optimizations because they can change the input vector's type and ordering to
+make them `UIntMappable`.
+
+If the input is not [`UIntMappable`](@ref), then we perform a presorted check and dispatch
+to [`QuickSort`](@ref).
+
+Otherwise, we dispatch to [`InsertionSort`](@ref) for inputs with `length <= 40` and then
+perform a presorted check ([`CheckSorted`](@ref)).
+
+We check for short inputs before performing the presorted check to avoid the overhead of the
+check for small inputs. Because the alternate dispatch is to [`InseritonSort`](@ref) which
+has efficient `O(n)` runtime on presorted inputs, the check is not necessary for small
+inputs.
+
+We check if the input is reverse-sorted for long vectors (more than 500 elements) because
+the check is essentially free unless the input is almost entirely reverse sorted.
+
+Note that once the input is determined to be [`UIntMappable`](@ref), we know the order forms
+a [total order](wikipedia.org/wiki/Total_order) over the inputs and so it is impossible to
+perform an unstable sort because no two elements can compare equal unless they _are_ equal,
+in which case switching them is undetectable. We utilize this fact to perform a more
+aggressive reverse sorted check that will reverse the vector `[3, 2, 2, 1]`.
+
+After these potential fast-paths are tried and failed, we [`ComputeExtrema`](@ref) of the
+input. This computation has a fairly fast `O(n)` runtime, but we still try to delay it until
+it is necessary.
+
+Next, we [`ConsiderCountingSort`](@ref). If the range the input is small compared to its
+length, we apply [`CountingSort`](@ref).
+
+Next, we [`ConsiderRadixSort`](@ref). This is similar to the dispatch to counting sort,
+but we conside rthe number of _bits_ in the range, rather than the range itself.
+Consequently, we apply [`RadixSort`](@ref) for any reasonably long inputs that reach this
+stage.
+
+Finally, if the input has length less than 80, we dispatch to [`InsertionSort`](@ref) and
+otherwise we dispatch to [`QuickSort`](@ref).
+"""
+const DEFAULT_STABLE = InitialOptimizations(IsUIntMappable(
```

Comment:
This seems quite hard to read (and modify). Would it be possible to simplify it for the reader, smth like as follows?
``` julia
const DEFAULT_STABLE = begin
          alg1 = QuickSort |> Small{80} |> ConsiderRadixSort |> ConsiderCountingSort |> ComputeExtrema |> CheckSorted |> Small{40} 
          IsUIntMappable(alg1, StableCheckSorted(QuickSort)) |> InitialOptimizations
       end
```
Also the outputed algorithm is kind of unusefull for the user :-D
``` julia
julia> Base.Sort.defalg([1,2])
Base.Sort.MissingOptimization{Base.Sort.BoolOptimization{Small{10, Base.Sort.InsertionSortAlg, Base.Sort.IEEEFloatOptimization{Base.Sort.IsUIntMappable{Small{40, Base.Sort.InsertionSortAlg, Base.Sort.CheckSorted{Base.Sort.ComputeExtrema{Base.Sort.ConsiderCountingSort{Base.Sort.CountingSort, ConsiderRadixSort{Base.Sort.RadixSort, Small{80, Base.Sort.InsertionSortAlg, PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}}}}, Base.Sort.StableCheckSorted{PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}}}}(Base.Sort.BoolOptimization{Small{10, Base.Sort.InsertionSortAlg, Base.Sort.IEEEFloatOptimization{Base.Sort.IsUIntMappable{Small{40, Base.Sort.InsertionSortAlg, Base.Sort.CheckSorted{Base.Sort.ComputeExtrema{Base.Sort.ConsiderCountingSort{Base.Sort.CountingSort, ConsiderRadixSort{Base.Sort.RadixSort, Small{80, Base.Sort.InsertionSortAlg, PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}}}}, Base.Sort.StableCheckSorted{PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}}}(Small{10, Base.Sort.InsertionSortAlg, Base.Sort.IEEEFloatOptimization{Base.Sort.IsUIntMappable{Small{40, Base.Sort.InsertionSortAlg, Base.Sort.CheckSorted{Base.Sort.ComputeExtrema{Base.Sort.ConsiderCountingSort{Base.Sort.CountingSort, ConsiderRadixSort{Base.Sort.RadixSort, Small{80, Base.Sort.InsertionSortAlg, PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}}}}, Base.Sort.StableCheckSorted{PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}}(Base.Sort.InsertionSortAlg(), Base.Sort.IEEEFloatOptimization{Base.Sort.IsUIntMappable{Small{40, Base.Sort.InsertionSortAlg, Base.Sort.CheckSorted{Base.Sort.ComputeExtrema{Base.Sort.ConsiderCountingSort{Base.Sort.CountingSort, ConsiderRadixSort{Base.Sort.RadixSort, Small{80, Base.Sort.InsertionSortAlg, PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}}}}, Base.Sort.StableCheckSorted{PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}(Base.Sort.IsUIntMappable{Small{40, Base.Sort.InsertionSortAlg, Base.Sort.CheckSorted{Base.Sort.ComputeExtrema{Base.Sort.ConsiderCountingSort{Base.Sort.CountingSort, ConsiderRadixSort{Base.Sort.RadixSort, Small{80, Base.Sort.InsertionSortAlg, PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}}}}, Base.Sort.StableCheckSorted{PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}(Small{40, Base.Sort.InsertionSortAlg, Base.Sort.CheckSorted{Base.Sort.ComputeExtrema{Base.Sort.ConsiderCountingSort{Base.Sort.CountingSort, ConsiderRadixSort{Base.Sort.RadixSort, Small{80, Base.Sort.InsertionSortAlg, PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}}}}(Base.Sort.InsertionSortAlg(), Base.Sort.CheckSorted{Base.Sort.ComputeExtrema{Base.Sort.ConsiderCountingSort{Base.Sort.CountingSort, ConsiderRadixSort{Base.Sort.RadixSort, Small{80, Base.Sort.InsertionSortAlg, PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}}}(Base.Sort.ComputeExtrema{Base.Sort.ConsiderCountingSort{Base.Sort.CountingSort, ConsiderRadixSort{Base.Sort.RadixSort, Small{80, Base.Sort.InsertionSortAlg, PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}}(Base.Sort.ConsiderCountingSort{Base.Sort.CountingSort, ConsiderRadixSort{Base.Sort.RadixSort, Small{80, Base.Sort.InsertionSortAlg, PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}(Base.Sort.CountingSort(), ConsiderRadixSort{Base.Sort.RadixSort, Small{80, Base.Sort.InsertionSortAlg, PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}(Base.Sort.RadixSort(), Small{80, Base.Sort.InsertionSortAlg, PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}(Base.Sort.InsertionSortAlg(), PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}(missing, missing, Base.Sort.InsertionSortAlg()))))))), Base.Sort.StableCheckSorted{PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}(PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}(missing, missing, Base.Sort.InsertionSortAlg())))))))
```

Reply:
Yeah... that output is especially terrible.

Reply:
I prefer not to reverse the order with `|>` because the standard order reflects the order of operations. `MissingOptimization(InsertionSort())` first tries to perform a missing optimization and then performs insertion sort.

I've reformatted the const declarations and added pretty printing:

```julia
const DEFAULT_STABLE = InitialOptimizations(
    IsUIntMappable(
        Small{40}(
            CheckSorted(
                ComputeExtrema(
                    ConsiderCountingSort(
                        ConsiderRadixSort(
                            Small{80}(
                                QuickSort)))))),
        StableCheckSorted(
            QuickSort)))
```
and
```
julia> Base.DEFAULT_STABLE
Base.Sort.MissingOptimization(
    Base.Sort.BoolOptimization(
        Base.Sort.Small{10}(
            InsertionSort(),
            Base.Sort.IEEEFloatOptimization(
                Base.Sort.IsUIntMappable(
                    Base.Sort.Small{40}(
                        InsertionSort(),
                        Base.Sort.CheckSorted(
                            Base.Sort.ComputeExtrema(
                                Base.Sort.ConsiderCountingSort(
                                    Base.Sort.CountingSort(),
                                    Base.Sort.ConsiderRadixSort(
                                        Base.Sort.RadixSort(),
                                        Base.Sort.Small{80}(
                                            InsertionSort(),
                                            PartialQuickSort(missing, missing,
                                                InsertionSort()))))))),
                    Base.Sort.StableCheckSorted(
                        PartialQuickSort(missing, missing,
                            InsertionSort())))))))
```
It's still fairly large, but I don't think there's any way of getting around that without either simplifying dispatch (possible a good idea for a future PR) or moving many optimization passes outside the scope of `DEFAULT_STABLE` (and therefore not easily configured by a user)

Reply:
> I've reformatted the const declarations and added pretty printing:

Thanks, it's definitely better. Btw, this seems like a nice case for reversed `<|` operator.